node/narwhal/src/helpers/codec.rs

```diff
@@ -84,3 +92,244 @@ impl<N: Network> Decoder for EventCodec<N> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Handshake(..) => unreachable!(),
+            Self::PostHandshake(ph_state) => Self::PostHandshake(ph_state.clone()),
+        }
+    }
+}
+
+impl NoiseState {
+    pub fn into_post_handshake_state(self) -> Self {
+        if let Self::Handshake(noise_state) = self {
+            let noise_state = noise_state.into_stateless_transport_mode().expect("handshake isn't finished");
+            Self::PostHandshake(PostHandshakeState { state: Arc::new(noise_state), tx_nonce: 0, rx_nonce: 0 })
+        } else {
+            unreachable!()
+        }
+    }
+}
+
+pub struct NoiseCodec<N: Network> {
+    codec: LengthDelimitedCodec,
+    event_codec: EventCodec<N>,
+    pub noise_state: NoiseState,
+}
+
+impl<N: Network> NoiseCodec<N> {
+    pub fn new(noise_state: NoiseState) -> Self {
+        Self { codec: LengthDelimitedCodec::new(), event_codec: EventCodec::default(), noise_state }
+    }
+}
+
+impl<N: Network> Encoder<EventOrBytes<N>> for NoiseCodec<N> {
+    type Error = std::io::Error;
+
+    fn encode(&mut self, message_or_bytes: EventOrBytes<N>, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let ciphertext = match self.noise_state {
+            NoiseState::Handshake(ref mut noise) => {
+                match message_or_bytes {
+                    // Don't allow message sending before the noise handshake has completed.
+                    EventOrBytes::Event(_) => unimplemented!(),
+                    EventOrBytes::Bytes(bytes) => {
+                        let mut buffer = [0u8; MAX_MESSAGE_LEN];
+                        let len = noise
+                            .write_message(&bytes, &mut buffer[..])
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        buffer[..len].into()
+                    }
+                }
+            }
+
+            NoiseState::PostHandshake(ref mut noise) => {
+                // Encode the message using the event codec.
+                let mut bytes = BytesMut::new();
+                match message_or_bytes {
+                    // Don't allow sending raw bytes after the noise handshake has completed.
+                    EventOrBytes::Bytes(_) => panic!("Unsupported post-handshake"),
+                    EventOrBytes::Event(event) => self.event_codec.encode(event, &mut bytes)?,
+                }
+
+                // Chunk the payload if necessary and encrypt with Noise.
+                //
+                // A Noise transport message is simply an AEAD ciphertext that is less than or
+                // equal to 65535 bytes in length, and that consists of an encrypted payload plus
+                // 16 bytes of authentication data.
+                //
+                // See: https://noiseprotocol.org/noise.html#the-handshakestate-object
+                let encrypted_chunks = bytes
+                    .par_chunks(MAX_MESSAGE_LEN - 16)
```

Comment:
```suggestion
                    .par_chunks(MAX_MESSAGE_LEN - TAG_LEN)
```
we should add this `const` so that it's clear where the `16` comes from